### ISSUE TITLE
Don't collect backtraces for try/catch blocks that never look at them

### ISSUE THREAD

**Issue by Keno** [JuliaLang/julia#43608]
For catch blocks that never look at the backtrace or rethrow, we should annotate in the exception frame that no unwinding needs to happen. In C code, we could do this manually with some sort of `JL_TRY_NO_BT` macro. In Julia code, ideally the optimizer could check whether the backtrace is required and if not make the appropriate annotation in the exception frame.

**Comment by vtjnash** [comment]
We nearly always leak this globally (via threads and/or the exception stack), so there are potentially no cases in which we can prove that the backtrace is not required.

**Comment by c42f** [comment]
Yeah this seems hard, I don't think we can do it in a neat way without language changes.

A big problem is that all our APIs for backtrace access are dynamically scoped, rather than a local syntactic construct within the catch block. So any function called from within `catch` could potentially call `catch_backtrace()` (or get the exception stack with `current_exceptions()`). To make progress here we'd need to track which functions call `catch_backtrace()` and prove that a given `catch` block doesn't call any such functions transitively.

The exception stack itself isn't a problem: if you could prove that an innermost `catch` never accesses the backtrace, you can avoid ever pushing that backtrace onto the exception stack.
